[DO NOT MERGE] kubelet: Refactor kuberuntime.SyncPod().

yifan-gu opened the pull request:
This splits the SyncPod into 3 parts:
- (1) Prune init containers
- (2) Compute changes (sandbox to kill, containers to kill,
    sandbox to start, containers to start).
- (3) Apply the changes (kill containers, kill sandboxes,
    start sandbox, start containers).

The change computation is further splited into 3 phases:
- (1) Construct list of sandboxes and containers that needs to
    be killed.
- (2) Construct list of sandbox and containers that needs to
    be started.
- (3) Make sure the sandbox will be killed if all active containers
    exited or are going to be killed.

In order to make the process more readable, a new type called
`syncAction` is added, which preserves the result of the computation
through different phases, and also provides functions to do operations
like container killing/starting based on the result.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34282)

<!-- Reviewable:end -->

Commit:
kubelet: Refactor kuberuntime.SyncPod().

This splits the SyncPod into 3 parts:

- (1) Prune init containers
- (2) Compute changes (sandbox to kill, containers to kill,
      sandbox to start, containers to start).
- (3) Apply the changes (kill containers, kill sandboxes,
      start sandbox, start containers).

The change computation is further splited into 3 phases:

- (1) Construct list of sandboxes and containers that needs to
      be killed.
- (2) Construct list of sandbox and containers that needs to
      be started.
- (3) Make sure the sandbox will be killed if all active containers
      exited or are going to be killed.

In order to make the process more readable, a new type called
`syncAction` is added, which preserves the result of the computation
through different phases, and also provides functions to do operations
like container killing/starting based on the result.

yifan-gu commented:
This is a fairly big PR, not expecting this to be merged one time, probably needs to split it into 2-3 PRs, but this PR gives the whole picture of what the result will look like, as well as a place to collect feedbacks.

/cc @kubernetes/sig-node @kubernetes/sig-rktnetes

yujuhong commented:
@freehan, FYI, @yifan-gu is trying to refactor the kuberuntime code to make it more readable. Since you're going to tackle the networking soon, there will be some unavoidable rebase
.

yujuhong commented:
@yifan-gu ping me when the PR is ready. Thanks!

Commit:
Rmove old code.

yifan-gu commented:
@k8s-bot cri test this.

yifan-gu commented:
@k8s-bot cri test this.

yifan-gu commented:
@yujuhong I tested locally to launch a pod with init containers, normal containers with this PR. Randomly kills sandbox, init containers, normal containers. Seems fine after I fixed some bugs.

So I think I can start testing this PR, and also split it into small PRs to ease the review.

yifan-gu commented:
@k8s-bot cri test this.

yifan-gu commented:
@k8s-bot cri node e2e test this

yifan-gu commented:
@k8s-bot cri node e2e test this

yifan-gu commented:
@k8s-bot cri node e2e test this

yifan-gu commented:
@yujuhong @euank @feiskyer Looks like the CRI tests are green with this PR. So next I will split this into 2-3 PRs for review (also update the unit tests along the way)

yujuhong commented:
> @yujuhong @euank @feiskyer Looks like the CRI tests are green now with this PR. So next I will split this into 2-3 PRs for review (also update the unit tests along the way)

@yifan-gu thanks! 

/cc @Random-Liu, who is also working on some unit test changes in the same package.

yifan-gu commented:
OK, rebased. This PR changes the kubelet to actually use new code path (which will is split to smaller PRs)

Let's see if this pass through tests.

Commit:
update BUILD

Commit:
Delete setupNetwork/teardownNetwork in syncpod

Also symlink the legacy container log files.

yifan-gu commented:
OK, fixed a bunch of node e2e failures by removing the code that does "network setup/teardown" in the kubelet.
4 failures left, probably related to the container log dir changes I need to rebase on, will fix tomorrow.

yifan-gu commented:
@k8s-bot bazel test this

yifan-gu commented:
OK, node e2e for CRI is passing. Debugging the gce e2e for CRI.

Commit:
add some debug logging.

Commit:
fix

Commit:
fill finished at field.

Commit:
rebase

yifan-gu commented:
Seems the `pull-kubernetes-e2e-gce-cri` is passing as well. All tests are passed (https://k8s-gubernator.appspot.com/pr/34282), although I have some issues kube-up with this PR. Still debugging.
But anyway this proves the refactoring is effective @yujuhong @Random-Liu @dchen1107 @feiskyer

feiskyer commented:
Cool. Will take a look later today.

yifan-gu commented:
@feiskyer Thanks, the first part is still in https://github.com/kubernetes/kubernetes/pull/35245 , nothing has been changed since last review.

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -752,3 +754,121 @@ func (m *kubeGenericRuntimeManager) removeContainerLog(containerID string) error
+	}
+
+	for _, id := range idsToPrune.UnsortedList() {
+		if err := m.runtimeService.RemoveContainer(id); err != nil {
```
Also clear ref for those containers?

yifan-gu replied:
good catch

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -541,7 +541,7 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
 //  5. Create init containers.
 //  6. Create normal containers.
-func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, _ v1.PodStatus, podStatus *kubecontainer.PodStatus, pullSecrets []v1.Secret, backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult) {
+func (m *kubeGenericRuntimeManager) OldSyncPod(pod *v1.Pod, _ v1.PodStatus, podStatus *kubecontainer.PodStatus, pullSecrets []v1.Secret, backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult) {
```
Remove OldSyncPod?

yifan-gu replied:
@feiskyer Yeah, just keep here for easier review (otherwise will remove a lot functions and creates a big diff). This PR is not expected to be merged, it's just for reference.

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -903,3 +883,73 @@ func (m *kubeGenericRuntimeManager) UpdatePodCIDR(podCIDR string) error {
 		})
 }
+
+// getSandboxInfoFromStatus returns the sandbox's ID, IP and sandbox config.
```
s/getSandboxInfoFromStatus/getSandboxInfoAndConfig

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -264,3 +264,43 @@ func (m *kubeGenericRuntimeManager) GetPortForward(podName, podNamespace string,
+}
+
+// runPodSandbox launches the pod sandbox.
+// On success, it returns the sandbox ID and the sandbox config.
```
nit: remove `and the sandbox config`

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -264,3 +264,43 @@ func (m *kubeGenericRuntimeManager) GetPortForward(podName, podNamespace string,
+
+	podSandBoxID, err := m.runtimeService.RunPodSandbox(podSandboxConfig)
+	if err != nil {
+		err := fmt.Errorf("CreatePodSandbox for pod %q failed, %v", format.Pod(pod), err)
```
s/CreatePodSandbox/RunPodSandbox

feiskyer reviewed (commented):
Hmm, needs more time to think about new sync action.

Commit:
address comments.

Commit:
BUILD

cblecker commented:
/remove-lifecycle frozen

vishh commented:
@yifan-gu closing this PR for now. If you are still pursuing this goal, kindly re-open.
